lib/controller/controller.go

```diff
@@ -950,9 +950,27 @@ func (ctrl *ProvisionController) provisionClaimOperation(claim *v1.PersistentVol
 			ctrl.eventRecorder.Event(claim, v1.EventTypeWarning, "ProvisioningCleanupFailed", strerr)
 		}
 	} else {
-		glog.Infof("volume %q provisioned for claim %q", volume.Name, claimToClaimKey(claim))
-		msg := fmt.Sprintf("Successfully provisioned volume %s", volume.Name)
-		ctrl.eventRecorder.Event(claim, v1.EventTypeNormal, "ProvisioningSucceeded", msg)
+		expectedVolumeMode := claim.Spec.VolumeMode
+		provisionedVolumeMode := volume.Spec.VolumeMode
+		// Report error if volumeMode for claim is different from volumeMode for provisioned volume,
+		// except if default volumeMode(Filesystem) is expected for claim and volumeMode is not set for provisioned volume.
+		if expectedVolumeMode != provisionedVolumeMode &&
+			!(expectedVolumeMode != nil && *expectedVolumeMode == v1.PersistentVolumeFilesystem && provisionedVolumeMode == nil) {
+			var expected, provisioned v1.PersistentVolumeMode
+			if expectedVolumeMode != nil {
+				expected = *expectedVolumeMode
+			}
+			if provisionedVolumeMode != nil {
+				provisioned = *provisionedVolumeMode
+			}
+			glog.Infof("volume %q is provisioned, but not bound to claim %q", volume.Name, claimToClaimKey(claim))
+			strerr := fmt.Sprintf("Volume %q is provisioned, but won't be bound to claim %q due to volumeMode mismatch(expected: %v, provisioned: %v)", volume.Name, claimToClaimKey(claim), expected, provisioned)
+			ctrl.eventRecorder.Event(claim, v1.EventTypeWarning, "ProvisioningFailed", strerr)
```

Comment:
Should the PV be deleted?  Will the external provisioner keep trying over and over?

Reply:
> Will the external provisioner keep trying over and over?
For as long as I see in my testing, no more volume is provisioned after one volume is provisioned.

> Should the PV be deleted? 
Ideally yes, I think. 

Let me check how it work after deleting the provisioned PV here and I will implement it and report you back.
My concerns currently are below two points:
1. Whether provision of the volume happen again after deleting the PV here, as you pointed out.
2. Both internal and external provisioner can be work in the same way.
> Will the external provisioner keep trying over and over?
For as long as I see in my testing, no more volume is provisioned after one volume is provisioned.

> Should the PV be deleted? 
Ideally yes, I think. 

Let me check how it work after deleting the provisioned PV here and I will report you back.
My concerns currently are below two points:
1. Whether provision of the volume happen again after deleting the PV here, as you pointed out.
2. Both internal and external provisioner can be work in the same way.

Reply:
As a result of some tests, I found that a new volume continues to be provisioned if the old volume is deleted, and it continues over and over. So, it would be better not to delete it, until some check mechanism is ready in provisioner to skip provisioning. (Please see the URL below for detail.)
Also, I found that it would be better to put everything in k8s side of the code, not external provisioner side. Because, if we put the fix to external provisioner side, this won't be enabled until provisioner is updated. As a result, old provisioner without this code continues to have the same problem, even when it is used with the latest version of the k8s.

Please see the k8s side of the fix below for the idea of the fix in detail. It will solve both internal provisioner case and external provisioner case.
https://github.com/kubernetes/kubernetes/pull/59947